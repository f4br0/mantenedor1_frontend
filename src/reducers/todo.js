export default function reducer(state, action) {
  switch (action.type) {
    case "ADD_TODO":
      if (!action.payload) {
        return state;
      }
      return {
        ...state,
        todos: [...state.todos, action.payload]
      };
    case "COMPLETE":
      return {
        ...state,
        todos: state.todos.map(t => {
          if (t.id === action.payload.id) {
            t.isActive = !t.isActive;
          }
          return t;
        })
      };
    case "SET_TODO":
      return {
        ...state,
        todos: action.payload
      };
    case "DELETE":
      return {
        ...state,
        todos: state.todos.filter(t => t.id !== action.payload.id)
      };
    default:
      return state;
  }
}
