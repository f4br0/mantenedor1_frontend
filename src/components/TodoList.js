import React, { useContext, useEffect, useState, useCallback } from "react";
import { Todo } from "../context/todo";
import TodoHeader from "./TodoHeader";
import TodoConsumer from "../api/TodoConsumer";

const TodoList = () => {
  const { state, dispatch } = useContext(Todo);
  const [isResult, setIsResult] = useState(null);

  function handleComplete(item) {
    const data = { ...item, isActive: !item.isActive }
    TodoConsumer.update(data, (status) => {

      dispatch({ type: "COMPLETE", payload: status.item });
    })
  }

  function deleteTask(item) {
    TodoConsumer.delete(item, (status) => {

      dispatch({ type: "DELETE", payload: item });
    })
  }


  useEffect(() => {
    TodoConsumer.getAll((status) => {
      setIsResult(status.isResult);
      if (status.isResult === 'success') {
        dispatch({ type: "SET_TODO", payload: status.data });
      }
    })
  }, [state.todo]);

  return <>{
    {
      success: <List handleComplete={handleComplete} state={state} deleteTask={deleteTask} />,
      error: <div className="alert alert-dark">Error in the request</div>,
      null: <div>Loading</div>
    }[isResult]
  }</>;
}

function heander(state) {
  const pluralize = count =>
    count > 1 ? `There are ${count} todos.` : `There is ${count} todo.`;

  return state.todos.length === 0 ? (
    <h4>Yay! All todos are done! Take a rest!</h4>
  ) : (
      <TodoHeader>
        <span className="float-right">{pluralize(state.todos.length)}</span>
      </TodoHeader>
    );
}

const List = ({ state, handleComplete, deleteTask }) => <div className="row">
  <div className="col-md-12">
    <div className="row">
      <div className="col-md-12">
        <br />
        {heander(state)}
      </div>
    </div>
    <div className="row">
      <div className="col-md-12">
        <ul className="list-group" data-testid="all-list">

          {
            state.todos.map(t => (
              <li key={t.id} className="list-group-item" >
                {`${t.description} | ${new Date(t.createdDate).toLocaleString()}`}
                <button
                  className="float-right btn btn-danger btn-sm"
                  style={{ marginLeft: 10 }}
                  onClick={() => deleteTask(t)}
                >
                  eliminar
                </button>
                <button
                  className="float-right btn btn-primary btn-sm"
                  style={{ marginLeft: 10 }}
                  onClick={() => handleComplete(t)}
                >
                  {t.isActive ? 'complete' : 'incomplete'}
                </button>


              </li>
            ))}
        </ul>
      </div>
    </div>
  </div>
</div>;

export default TodoList;
