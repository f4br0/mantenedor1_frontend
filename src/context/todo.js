import React, { useContext, useReducer } from "react";

import reducers from "../reducers/todo";

import TodoList from "../components/TodoList";
import TodoForm from "../components/TodoForm";

// Initial Data
export const Todo = React.createContext({
  todos: []
});

export default () => {
  const globalStore = useContext(Todo);
  const [state, dispatch] = useReducer(reducers, globalStore);
  return (
    <Todo.Provider value={{ state, dispatch }}>
      <TodoForm />
      <TodoList />
    </Todo.Provider>
  );
};

