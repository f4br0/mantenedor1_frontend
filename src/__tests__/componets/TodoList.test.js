import React from "react";
import { render, fireEvent, waitForElement } from 'react-testing-library'
import { Todo } from "../../context/todo";
import TodoConsumer from "../../api/TodoConsumer";
import TodoList from "../../components/TodoList";

test("<TodoList /> #display",  () => {

  const todos = [
    { item: "task 1", id: 1 },
    { item: "task 2", id: 2 },
    { item: "task 3", id: 3 }
  ];
  const dispatch = jest.fn();//mock validar el dispatch

  //dependencia que consume la API
  jest.spyOn(TodoConsumer, "getAll");
  TodoConsumer.getAll = jest.fn(cb => cb({ data: todos, isResult: 'success' }));

  //renderiza el contenedor
  const { container, getByTestId } = render(
    <Todo.Provider value={{ state: { todos: todos }, dispatch }}>
      <TodoList />
    </Todo.Provider>
  );

  expect(container.getElementsByTagName("li").length).toBe(3);
  //caputar el html de test ID y lo guarda
  expect(getByTestId('all-list').innerHTML).toMatchSnapshot();
  
  expect(TodoConsumer.getAll).toHaveBeenCalledTimes(1);
  expect(dispatch).toHaveBeenCalledWith({ type: "SET_TODO", payload: todos });
});

test("<TodoList /> #completeCalls", () => {

  const dispatch = jest.fn();//mock validar el dispatch
  const todos = [
    { item: "task 1", id: 1 },
    { item: "task 2", id: 2 },
    { item: "task 3", id: 3 }
  ];

  jest.spyOn(TodoConsumer, "update");
  TodoConsumer.update = jest.fn((data, cb) => cb({ item: {id: 200}}));

  const { container } = render(
    <Todo.Provider value={{ state: { todos }, dispatch }}>
      <TodoList />
    </Todo.Provider>
  );

  for (let element of container.querySelectorAll("li > button")) {
    fireEvent.click(element);
  }

  expect(TodoConsumer.update.mock.calls.length).toBe(3);
});

test("<TodoList /> #allTodosAreDone", async () => {
  const state = { todos: [] };
  const dispatch = jest.fn();//mock validar el dispatch
 
  const { queryByText } = render(
    <Todo.Provider value={{ state , dispatch}}>
      <TodoList />
    </Todo.Provider>
  );
  expect(queryByText(/Yay! All todos are done!/i).hasChildNodes()).toBe(true)
});

test("<TodoList /> #oneITem",  () => {
  const state = { todos: [ { item: "task 1", id: 1 }] };
  const dispatch = jest.fn();//mock validar el dispatch
  const { queryByText } = render(
    <Todo.Provider value={{ state, dispatch }}>
      <TodoList />
    </Todo.Provider>
  );
  expect(queryByText(/There is/i).hasChildNodes()).toBe(true)
});
