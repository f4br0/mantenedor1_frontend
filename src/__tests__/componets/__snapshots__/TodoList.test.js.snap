// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> #display 1`] = `"<li class=\\"list-group-item\\"><button class=\\"float-right btn btn-danger btn-sm\\" style=\\"margin-left: 10px;\\">eliminar</button><button class=\\"float-right btn btn-primary btn-sm\\" style=\\"margin-left: 10px;\\">incomplete</button></li><li class=\\"list-group-item\\"><button class=\\"float-right btn btn-danger btn-sm\\" style=\\"margin-left: 10px;\\">eliminar</button><button class=\\"float-right btn btn-primary btn-sm\\" style=\\"margin-left: 10px;\\">incomplete</button></li><li class=\\"list-group-item\\"><button class=\\"float-right btn btn-danger btn-sm\\" style=\\"margin-left: 10px;\\">eliminar</button><button class=\\"float-right btn btn-primary btn-sm\\" style=\\"margin-left: 10px;\\">incomplete</button></li>"`;
